servicelayerfunctionalities.cc

```diff
@@ -0,0 +1,328 @@
+    message->set_username(request->username());
+    message->set_text(request->text());
+    message->set_id(nextChirpID);  
+    chirp::Timestamp* timestamp = message->mutable_timestamp();
+    timestamp->set_seconds(static_cast<int64_t>(seconds));
+    timestamp->set_useconds(microseconds_since_epoch);
+    message->set_parent_id(request->parent_id());
+    message->SerializeToString(&value);
+    
+    setChirpReply(message, response);
+  }
+
+  user.SerializeToString(&userKey);
+  clientKey.put(request->username(),userKey); //Add Chirp to chirp::User
+  clientKey.put("chirp"+nextChirpID,value);//Add chirp::Chirp to database
+  
+  //If its a reply to an existing chirp
+  if (request->parent_id() != "") { 
+    std::string value2; 
+    {
+      std::vector <std::string> fromget = clientKey.get("reply"+request->parent_id());
+
+      chirp::ChirpReplies replies;
+      
+      //If chirp id is not found, use the 'replies' as a new ChirpReplies and set parent_id()
+      if (fromget.size() == 0) { 
+        replies.set_parent_id(request->parent_id());
+      } else { //If found, parse from existing ChirpReplies
+        replies.ParseFromString(fromget[0]);
+      }
+      chirp::Chirp* reply = replies.add_chirps();
+      reply->set_username(request->username());
+      reply->set_text(request->text());
+      reply->set_id(nextChirpID);
+      reply->set_parent_id(request->parent_id());
+      chirp::Timestamp* timestamp = reply->mutable_timestamp();
+      timestamp->set_seconds(static_cast<int64_t>(seconds));
+      timestamp->set_useconds(microseconds_since_epoch);
+      replies.SerializeToString(&value2);
+    }
+    clientKey.put("reply"+request->parent_id(),value2); //Add reply<ID> to backend
+  }
+  
+  return Status::OK;
+}
+
+Status ServerForCommandLineClient::read(ServerContext* context, const ReadRequest* request, ReadReply* response) {
+  /*
+    Tested: Able to get all chirps and it's replies and send it back via ReadReply
+  */
+  ClientForKeyValueStore clientKey(grpc::CreateChannel("localhost:50000", grpc::InsecureChannelCredentials()));
+
+  std::vector <std::string> fromget = clientKey.get("chirp"+request->chirp_id());
+  std::vector <chirp::Chirp> chirpToSend; //For sorting by timestamp
+  if (fromget.size() == 0) {
+    return Status::CANCELLED;
+  }
+  std::string mainChirpMsg = fromget[0];
+
+  chirp::Chirp* mainChirp = response->add_chirps();
+  mainChirp->ParseFromString(mainChirpMsg);
+  chirpToSend.push_back(*mainChirp);
+
+  std::queue<chirp::Chirp> myqueue;
+  myqueue.push(*mainChirp);
+  
+  //Running DFS to get threads related to chirp<id>
+  while(!myqueue.empty()) {
+    chirp::Chirp currentChirp = myqueue.front();
+    myqueue.pop();
+    
+    //Find other chirps that are in reply to this currentChirp
+    fromget = clientKey.get("reply"+currentChirp.id());
+    if(fromget.size() == 0) {
+      continue;
+    }
+
+    chirp::ChirpReplies chirpReplies = convertToChirpReplies(fromget[0]);
```

Comment:
Variable names should be lowercase w/ underscores as appropriate

Reply:
done